### PR TITLE
feat: Add async deletion model

### PR DESCRIPTION
**Pull request by macobo**
The async deletion model will be used to schedule async deletions of
team/person/group data as not to disturb other operations in the
cluster.

See https://github.com/PostHog/posthog/issues/11128 for more details

This is not yet wired up for anything - want to get feedback on the core
backend logic first.

:facepalm: thought I PR-ed this yesterday but apparently not.

### CODE REVIEW

File: posthog/models/async_deletion/delete.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        "Starting AsyncDeletion for teams on other tables",
+        {"count": len(team_deletions), "team_ids": list(set(row.team_id for row in deletions)),},
+    )
+    conditions, args = _conditions(team_deletions)
+    for table in TABLES_TO_DELETE_TEAMS_FROM:
+        sync_execute(
+            f"""
+            ALTER TABLE {table}
+            ON CLUSTER '{CLICKHOUSE_CLUSTER}'
+            DELETE WHERE {" OR ".join(conditions)}
```
Comment (yakkomajuri) [inline posthog/models/async_deletion/delete.py:62]: q: say this query gets massively large - is that a problem for CH to handle? I assume no but just checking

File: posthog/models/async_deletion/async_deletion.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+            models.UniqueConstraint(
+                name="unique deletion for groups", fields=["deletion_type", "key", "group_type_index"],
+            ),
+        ]
+        indexes = [
+            models.Index(name="delete_verified_at index", fields=["delete_verified_at"]),
+        ]
+
+    id: models.BigAutoField = models.BigAutoField(primary_key=True)
+    # Should be one of the enum above
```
Comment (yakkomajuri) [inline posthog/models/async_deletion/async_deletion.py:30]:
```suggestion
    # Should be one of the DeletionType enum
```

File: posthog/models/async_deletion/async_deletion.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    deletion_type: models.PositiveSmallIntegerField = models.PositiveSmallIntegerField(null=False, blank=False)
+
+    team: models.ForeignKey = models.ForeignKey("Team", on_delete=models.CASCADE)
+
+    # id for team (same as team_id column), uuid for person, key for group
+    key: models.CharField = models.CharField(max_length=400, null=False, blank=False)
+    # Only populated for group deletions
+    group_type_index: models.IntegerField = models.IntegerField(null=True, blank=False)
```
Comment (yakkomajuri) [inline posthog/models/async_deletion/async_deletion.py:38]: wondering if it makes sense to have some sort of additional context JSON field that would be more extensible than this

File: posthog/models/test/test_async_deletion_model.py [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+            deletion_type=DeletionType.Team, team_id=self.teams[0].pk, key=str(self.teams[0].pk), created_by=self.user
+        )
+
+        mark_deletions_done()
+
+        deletion.refresh_from_db()
+        self.assertIsNotNone(deletion.delete_verified_at)
+
+    @snapshot_clickhouse_queries
+    def test_mark_team_deletions_done_when_not_done(self):
```
Comment (yakkomajuri) [inline posthog/models/test/test_async_deletion_model.py:46]: following the pattern you use below

```suggestion
    def test_mark_deletions_done_team_when_not_done(self):
```

File: posthog/models/async_deletion/async_deletion.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+    created_by: models.ForeignKey = models.ForeignKey("User", on_delete=models.CASCADE)
```
Comment (yakkomajuri) [inline posthog/models/async_deletion/async_deletion.py:40]: do we want cascade here? say a user is gone we probably still want to process the issued delete? maybe SET_NULL to remove references to the user?

File: posthog/models/async_deletion/delete.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+def _fetch_unverified_deletions_grouped():
+    result = defaultdict(list)
+    for item in AsyncDeletion.objects.filter(delete_verified_at__isnull=True):
+        key = (item.deletion_type, item.group_type_index)
+        result[key].append(item)
+    return result
+
+
+def _verify_by_group(deletion_type: int, async_deletions: List[AsyncDeletion]) -> List[AsyncDeletion]:
```
Comment (yakkomajuri) [inline posthog/models/async_deletion/delete.py:94]: nit: I was briefly confused on a reference by the name of this function given group is also a deletion type

**Review by yakkomajuri — COMMENTED** [review]
approach seems reasonable to me, some initial comments

File: posthog/models/async_deletion/async_deletion.py (hunk 3 above)
Comment (macobo) [reply posthog/models/async_deletion/async_deletion.py:38]: If we go with a JSON field then we need to roll validation into the app. YAGNI - adding columns is cheap.

File: posthog/models/async_deletion/delete.py (hunk 1 above)
Comment (macobo) [reply posthog/models/async_deletion/delete.py:62]: Define massively large and problem to handle?

We're bottlenecked currently on rewriting $ALL_OF_DATA on disk and I don't see ~20 conditions suddenly making this CPU-bound. If there's query complexity issues we have Sentry for that.

File: posthog/models/__init__.py [hunk 7]
```diff
@@ -43,12 +44,14 @@
...
     "ActivityLog",
     "Annotation",
+    "AsyncDeletion",
     "AsyncMigration",
     "AsyncMigrationError",
     "Cohort",
     "CohortPeople",
     "Dashboard",
     "DashboardTile",
+    "DeletionType",
```
Comment (yakkomajuri) [inline posthog/models/__init__.py:54]: is this here just so we don't get complaints about it not being used anywhere?

**Review by yakkomajuri — APPROVED** [review]

File: posthog/models/async_deletion/async_deletion.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    # id for team (same as team_id column), uuid for person, key for group
+    key: models.CharField = models.CharField(max_length=400, null=False, blank=False)
+    # Only populated for group deletions
+    group_type_index: models.IntegerField = models.IntegerField(null=True, blank=False)
+
+    created_by: models.ForeignKey = models.ForeignKey("User", null=True, on_delete=models.SET_NULL)
+    created_at: models.DateTimeField = models.DateTimeField(auto_now_add=True)
+
+    # When was the data verified to be deleted - we can skip it in the next round
+    delete_verified_at: models.DateTimeField = models.DateTimeField(null=True, blank=True)
```
[the remark this answers is not in the document]
Comment (macobo) [reply posthog/models/async_deletion/async_deletion.py:44]: I imagine we keep them for audit purposes.

File: posthog/models/async_deletion/delete.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    "person_distinct_id",
+    "person_distinct_id2",
+    "groups",
+    "cohortpeople",
+    "person_static_cohort",
+    "plugin_log_entries",
+]
+
+
+def run_event_table_deletions():
```
[the remark this answers is not in the document]
Comment (macobo) [reply posthog/models/async_deletion/delete.py:24]: This rename gets me to a rebase hell so leaving this for now.

**Pull request merged**